File: backend/src/models/Farm.ts
import mongoose, { Document, Schema, Model } from 'mongoose';

// Interface for sensor readings
interface ISensorReadings {
  soilMoisture?: number;
  temperature?: number;
  humidity?: number;
  lightIntensity?: number;
  soilPh?: number;
}

// Interface for sensor
interface ISensor {
  deviceID: string;
  sensorName: string;
  description: string;
  soilType: 'loamy' | 'sandy' | 'clay' | 'silty';
  readings: ISensorReadings;
}

// Interface for field
interface IField {
  fieldName: string;
  plantingDate: Date;
  growthStage: 'VE' | 'V3' | 'V8' | 'VT' | 'R1' | 'R6';
  sensors: ISensor[];
}

// Interface for static methods
interface IFarmModel extends Model<IFarm> {
  getStatistics(): Promise<any>;
}

export interface IFarm extends Document {
  userId: mongoose.Types.ObjectId;
  farmName: string;
  location: string;
  fields: IField[];
  images?: string[];
  createdAt: Date;
  updatedAt: Date;
  
  // Instance methods
  addField(fieldData: IField): Promise<IFarm>;
  updateField(fieldName: string, fieldData: Partial<IField>): Promise<IFarm>;
  removeField(fieldName: string): Promise<IFarm>;
  addSensorToField(fieldName: string, sensorData: ISensor): Promise<IFarm>;
  removeSensorFromField(fieldName: string, deviceID: string): Promise<IFarm>;
  updateGrowthStages(): Promise<IFarm>;
}

const farmSchema = new Schema<IFarm>(
  {
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    farmName: {
      type: String,
      required: true
    },
    location: {
      type: String,
      required: true
    },
    fields: [
      {
        fieldName: {
          type: String,
          required: true
        },
        plantingDate: {
          type: Date,
          required: true
        },
        growthStage: {
          type: String,
          enum: ['VE', 'V3', 'V8', 'VT', 'R1', 'R6'],
          default: 'VE'
        },
        sensors: [
          {
            deviceID: {
              type: String,
              required: true
            },
            sensorName: {
              type: String,
              required: true
            },
            description: {
              type: String,
              required: true
            },
            soilType: {
              type: String,
              enum: ['loamy', 'sandy', 'clay', 'silty'],
              required: true
            },
            readings: {
              soilMoisture: {
                type: Number,
                default: 0
              },
              temperature: {
                type: Number,
                default: 0
              },
              humidity: {
                type: Number,
                default: 0
              },
              lightIntensity: {
                type: Number,
                default: 0
              },
              soilPh: {
                type: Number,
                default: 0
              }
            }
          }
        ]
      }
    ]
  },
  { 
    timestamps: true,
    collection: 'farms'
  }
);

// Compound indexes for performance
farmSchema.index({ userId: 1 });
farmSchema.index({ farmName: 1, userId: 1 }, { unique: true });
farmSchema.index({ 'fields.fieldName': 1 });
farmSchema.index({ 'fields.sensors.deviceID': 1 });
farmSchema.index({ 'fields.growthStage': 1 });

// Pre-save middleware to auto-update growth stages based on planting dates
farmSchema.pre('save', function(next) {
  this.fields.forEach((field: any) => {
    if (field.plantingDate) {
      const now = new Date();
      const diffTime = Math.abs(now.getTime() - field.plantingDate.getTime());
      const daysSincePlanting = Math.ceil(diffTime / (1000 * 60 * 60 * 24));

      if (daysSincePlanting <= 7) field.growthStage = 'VE'; // Emergence (0-7 days)
      else if (daysSincePlanting <= 21) field.growthStage = 'V3'; // 3rd leaf (8-21 days)
      else if (daysSincePlanting <= 42) field.growthStage = 'V8'; // 8th leaf (22-42 days)
      else if (daysSincePlanting <= 65) field.growthStage = 'VT'; // Tasseling (43-65 days)
      else if (daysSincePlanting <= 85) field.growthStage = 'R1'; // Silking (66-85 days)
      else field.growthStage = 'R6'; // Maturity (86+ days)
    }
  });
  next();
});

// Static method to get farm statistics
farmSchema.statics.getStatistics = async function() {
  const stats = await this.aggregate([
    {
      $group: {
        _id: null,
        totalFarms: { $sum: 1 },
        totalFields: { $sum: { $size: '$fields' } },
        farmsWithFields: {
          $sum: { $cond: [{ $gt: [{ $size: '$fields' }, 0] }, 1, 0] }
        },
        totalSensors: {
          $sum: {
            $sum: {
              $map: {
                input: '$fields',
                as: 'field',
                in: { $size: '$$field.sensors' }
              }
            }
          }
        }
      }
    }
  ]);

  return stats[0] || {
    totalFarms: 0,
    totalFields: 0,
    farmsWithFields: 0,
    totalSensors: 0
  };
};

// Instance method to add field
farmSchema.methods.addField = function(fieldData: IField) {
  this.fields.push(fieldData);
  return this.save();
};

// Instance method to update field
farmSchema.methods.updateField = function(fieldName: string, fieldData: Partial<IField>) {
  const fieldIndex = this.fields.findIndex((field: any) => field.fieldName === fieldName);
  if (fieldIndex !== -1) {
    Object.assign(this.fields[fieldIndex], fieldData);
  }
  return this.save();
};

// Instance method to remove field
farmSchema.methods.removeField = function(fieldName: string) {
  this.fields = this.fields.filter((field: any) => field.fieldName !== fieldName);
  return this.save();
};

// Instance method to add sensor to field
farmSchema.methods.addSensorToField = function(fieldName: string, sensorData: ISensor) {
  const field = this.fields.find((field: any) => field.fieldName === fieldName);
  if (field) {
    field.sensors.push(sensorData);
  }
  return this.save();
};

// Instance method to remove sensor from field
farmSchema.methods.removeSensorFromField = function(fieldName: string, deviceID: string) {
  const field = this.fields.find((field: any) => field.fieldName === fieldName);
  if (field) {
    field.sensors = field.sensors.filter((sensor: any) => sensor.deviceID !== deviceID);
  }
  return this.save();
};

// Instance method to update growth stages
farmSchema.methods.updateGrowthStages = function() {
  // Growth stages are automatically updated in pre-save middleware
  return this.save();
};

const Farm = mongoose.model<IFarm, IFarmModel>('Farm', farmSchema);

export default Farm;

File: frontend/mobile/lib/features/farm/domain/entities/farm.dart
import 'package:equatable/equatable.dart';

// Sensor readings model
class SensorReadings extends Equatable {
  final double soilMoisture;
  final double temperature;
  final double humidity;
  final double lightIntensity;
  final double soilPh;

  const SensorReadings({
    this.soilMoisture = 0,
    this.temperature = 0,
    this.humidity = 0,
    this.lightIntensity = 0,
    this.soilPh = 0,
  });

  Map<String, dynamic> toJson() {
    return {
      'soilMoisture': soilMoisture,
      'temperature': temperature,
      'humidity': humidity,
      'lightIntensity': lightIntensity,
      'soilPh': soilPh,
    };
  }

  factory SensorReadings.fromJson(Map<String, dynamic> json) {
    return SensorReadings(
      soilMoisture: (json['soilMoisture'] ?? 0).toDouble(),
      temperature: (json['temperature'] ?? 0).toDouble(),
      humidity: (json['humidity'] ?? 0).toDouble(),
      lightIntensity: (json['lightIntensity'] ?? 0).toDouble(),
      soilPh: (json['soilPh'] ?? 0).toDouble(),
    );
  }

  @override
  List<Object?> get props => [soilMoisture, temperature, humidity, lightIntensity, soilPh];
}

// Sensor model
class Sensor extends Equatable {
  final String deviceID;
  final String sensorName;
  final String description;
  final String soilType; // loamy, sandy, clay, silty
  final SensorReadings readings;

  const Sensor({
    required this.deviceID,
    required this.sensorName,
    required this.description,
    required this.soilType,
    required this.readings,
  });

  Map<String, dynamic> toJson() {
    return {
      'deviceID': deviceID,
      'sensorName': sensorName,
      'description': description,
      'soilType': soilType,
      'readings': readings.toJson(),
    };
  }

  factory Sensor.fromJson(Map<String, dynamic> json) {
    return Sensor(
      deviceID: json['deviceID'] ?? json['deviceId'] ?? json['_id'] ?? '',
      sensorName: json['sensorName'] ?? json['name'] ?? '',
      description: json['description'] ?? '',
      soilType: json['soilType'] ?? 'loamy',
      readings: SensorReadings.fromJson(json['readings'] ?? {}),
    );
  }

  @override
  List<Object?> get props => [deviceID, sensorName, description, soilType, readings];
}

// Field model
class Field extends Equatable {
  final String fieldName;
  final DateTime plantingDate;
  final String growthStage; // VE, V3, V8, VT, R1, R6
  final List<Sensor> sensors;

  const Field({
    required this.fieldName,
    required this.plantingDate,
    required this.growthStage,
    required this.sensors,
  });

  Map<String, dynamic> toJson() {
    return {
      'fieldName': fieldName,
      'plantingDate': plantingDate.toIso8601String(),
      'growthStage': growthStage,
      'sensors': sensors.map((sensor) => sensor.toJson()).toList(),
    };
  }

  factory Field.fromJson(Map<String, dynamic> json) {
    return Field(
      fieldName: json['fieldName'] ?? '',
      plantingDate: DateTime.parse(json['plantingDate']),
      growthStage: json['growthStage'] ?? 'VE',
      sensors: (json['sensors'] as List<dynamic>? ?? [])
          .map((sensorJson) => Sensor.fromJson(sensorJson))
          .toList(),
    );
  }

  @override
  List<Object?> get props => [fieldName, plantingDate, growthStage, sensors];
}

class Farm extends Equatable {
  final String? id;
  final String userId;
  final String farmName;
  final String location;
  final List<Field> fields;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Farm({
    this.id,
    required this.userId,
    required this.farmName,
    required this.location,
    required this.fields,
    required this.createdAt,
    required this.updatedAt,
  });

  Farm copyWith({
    String? id,
    String? userId,
    String? farmName,
    String? location,
    List<Field>? fields,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Farm(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      farmName: farmName ?? this.farmName,
      location: location ?? this.location,
      fields: fields ?? this.fields,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'farmName': farmName,
      'location': location,
      'fields': fields.map((field) => field.toJson()).toList(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  factory Farm.fromJson(Map<String, dynamic> json) {
    // Handle userId - it can be either a string or an object with _id
    String userIdValue = '';
    if (json['userId'] is String) {
      userIdValue = json['userId'];
    } else if (json['userId'] is Map<String, dynamic>) {
      userIdValue = json['userId']['_id'] ?? '';
    }

    return Farm(
      id: json['id'] ?? json['_id'],
      userId: userIdValue,
      farmName: json['farmName'] ?? '',
      location: json['location'] ?? '',
      fields: (json['fields'] as List<dynamic>? ?? [])
          .map((fieldJson) => Field.fromJson(fieldJson))
          .toList(),
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  @override
  List<Object?> get props => [
        id,
        userId,
        farmName,
        location,
        fields,
        createdAt,
        updatedAt,
      ];

  @override
  String toString() {
    return 'Farm('
        'id: $id, '
        'userId: $userId, '
        'farmName: $farmName, '
        'fields: ${fields.length}, '
        'createdAt: $createdAt, '
        'updatedAt: $updatedAt'
        ')';
  }
}